### PR TITLE
feat(win32): SSH host ingestion into profiles and palette (#143)

### LINKED ISSUE #143: C20 · Lean SSH host ingestion into profiles/palette (Wave 3)
**Wave 3 · cost M · ruling: adopt the lean slice only, per the amended benchmark-user scope ("…and the remote hosts they reach over SSH").**

From the [2026 Competitive Roadmap](https://github.com/amanthanvi/winghostty/blob/claude/winghostty-competitive-analysis-9ndxmv/plans/2026-competitive-roadmap.md) (entry C20), decided in the competitive-analysis [decision session](``https://github.com/amanthanvi/winghostty/blob/claude/winghostty-competitive-analysis-9ndxmv/plans/wayfinder/competitive-analysis/tickets/D01-decision-session.md).``

**Why:** the benchmark user's week includes remote hosts — Tabby's 74k stars are built on connection management; WezTerm auto-populates SSH domains; Windows Terminal added SSH auto-detection in 1.24. Discovery is the missing piece: winghostty already wraps ssh for terminfo.

**Shape:** parse `%USERPROFILE%\.ssh\config`, surface hosts as launchable profiles/palette entries running the system `ssh`. Explicitly out: bundled SSH client, secrets vault (deferred C24), fleet tools.

**Evidence:** synthesis C20; [tabby](https://github.com/amanthanvi/winghostty/blob/claude/winghostty-competitive-analysis-9ndxmv/plans/wayfinder/competitive-analysis/research/tabby.md) §5/§10.1, [wezterm](https://github.com/amanthanvi/winghostty/blob/claude/winghostty-competitive-analysis-9ndxmv/plans/wayfinder/competitive-analysis/research/wezterm.md) §5/§10.5, windows-terminal §3.

**Dependencies:** none; deferred C24 (Credential Manager secrets) waits on this proving out.

### PR DESCRIPTION
**Pull request by amanthanvi**
## Summary

Surfaces the user's own SSH hosts. `%USERPROFILE%\.ssh\config` is parsed for
concrete `Host` aliases, which appear as `SSH: <alias>` in the profile picker
and `Connect to <alias>` in the universal palette. Selecting one launches the
resolved system `ssh.exe` with the bare alias as its only argument, from the
user's home directory, so `ssh` reads its own configuration.

The lean C20 slice only: no bundled SSH client, no secrets vault (deferred
C24), no fleet or connection-management tooling.

Fixes #143

## Changes

- New `src/config/windows_ssh_hosts.zig`: bounded, pure-Zig `~/.ssh/config`
  parser. Only `Host` lines are read; every other keyword is left to `ssh`.
  Multiple aliases per line; `Match` blocks ignored, including their includes;
  comments, BOM, CRLF, `Key=Value`, and case-insensitive keywords tolerated;
  missing file yields an empty list, unreadable warns once. Aliases are
  deduplicated by exact byte equality, so case-distinct OpenSSH aliases remain distinct.
- `listProfiles` takes `include_ssh_hosts` and appends SSH entries after the
  detected shells, keyed `ssh:<alias>`, with `ssh.exe` resolved from PATH and
  falling back to `C:\Windows\System32\OpenSSH\ssh.exe`. SSH discovery is
  additive: any failure degrades to the detected shells rather than leaving the
  user with no picker.
- New `ProfileKind.ssh` with its own badge and glyph; palette rows get the
  "SSH host" subtitle through one optional owned `palette_title` on `Profile`.
- New `ssh-config-hosts` config knob (default `true`).
- Docs: `windows.md` "SSH hosts" section, `status.md` bullet, capability
  matrix row.

### Enforcement

SSH-specific hardening lives in one helper that every launch path calls — new
surface, split, and restore — so no path can inherit the command without the
guarantees attached to it:

- `shell-integration` is forced off. It is a local-shell mechanism that cannot
  reach a remote session, and leaving it on lets the integration rewrite the
  two-element ssh argv into a space-joined, unquoted `cmd.exe /C` string.
- The working directory resolves to `%USERPROFILE%` explicitly — the same
  directory `ssh` expands `~` to, and deliberately not `HOMEDRIVE`+`HOMEPATH`,
  which can point at a network share on a domain-joined machine.

An alias is one literal argv element. Aliases starting with `-` are refused so
they cannot become ssh options, as are shell/`cmd` metacharacters and any byte
outside printable ASCII.

Session restore refuses any `ssh:` profile key case-insensitively before exact
live-profile lookup rather than dialing out unattended at startup. A refused
pane also drops its profile key, so a later split cannot re-resolve the
connection. Whether a surface was launched over SSH is recorded from the
profile kind at launch rather than re-derived from the key.

Include handling is one level deep and capped at 16 files. Root and Include
opens run off the UI thread through a single-in-flight worker under one shared
50 ms refresh budget; UNC/device paths are rejected before opening.

### Incidental fixes

Two pre-existing defects found while working in this area:

- **Double free** in `safeCurrentDirectory`: both the caller and
  `safeCurrentDirectoryWithCurrent` freed `current` on the path where no
  explicit cwd was requested and a current directory existed. Ownership now
  sits entirely in the callee.
- **Uninitialized stack read** in `drawPaletteRowText`: it fell back to
  `buf.len` when `utf8ToUtf16Le` failed, rendering whatever was on the stack
  past the partial prefix. It now draws nothing.

Existing working-directory precedence is otherwise unchanged: SSH gets its
home directory through an explicit resolved path rather than by promoting the
`working-directory = home` flag, so non-SSH profiles (including WSL's `--cd ~`)
behave exactly as before, with a regression test.

## Validation

- `zig fmt --check src` — only the pre-existing failure on the untouched
  generated file `src/build/uucode_tables.zig`.
- `zig build -Demit-exe=true` — pass.
- `zig build test -Demit-test-exe=true` — **full suite green**, no failures.
- `zig build test -Dtest-filter=ssh` / `-Dtest-filter=profile`
  `-Dtest-filter=Profile` `-Dtest-filter=safeCurrentDirectory`
  `-Dtest-filter=session` `-Dtest-filter=palette` — pass.
- `pwsh -NoProfile -File scripts/check-source-format.ps1` — pass.
- Interactive (sandbox `USERPROFILE` with two hosts plus a `Host *` block):
  both aliases appear in the picker and palette; launching one runs
  `ssh.exe` + the bare alias with cwd = the sandbox home and the tab titled
  with the alias. Screenshots: `.t3\worktrees\winghostty\evidence\143\`.

New tests cover the case-variance restore bypass, the split-path invariant,
non-ASCII alias rejection, UNC/device include rejection, the include-count cap,
and both ssh.exe resolution failure modes.

## Residuals

- 256 SSH hosts can crowd theme entries out of the 768-slot palette catalog.
- The System32 fallback path is hardcoded rather than `%SystemRoot%`-derived.
- One-shot warn flags mean a later, different discovery failure is silent.
- The interactive evidence predates the shell-integration, session-restore, and
  `%USERPROFILE%` changes; those are covered by unit tests and build only.

Follow-up prompt (phased, commit per phase):
`.t3\worktrees\winghostty\prompts\c4-143-followup-PARKED.md`.

## Summary by Sourcery

Add safe, configurable Windows SSH host discovery and one-click profile and palette entries backed by the system SSH client.

New Features:
- Discover concrete aliases from the Windows user's SSH configuration and expose them in the profile picker and universal palette.
- Launch selected SSH hosts through the system ssh.exe with the alias as a literal argument and SSH-specific profile presentation.

Bug Fixes:
- Prevent SSH profiles from being relaunched during session restore or unintentionally inherited during splits.
- Fix ownership handling in safeCurrentDirectory to eliminate a double free.
- Avoid rendering uninitialized stack data when palette text conversion fails.

Enhancements:
- Harden SSH launches with explicit user-home working directories, disabled shell integration, bounded alias validation, and safe handling of SSH profile identity and titles.
- Bound and isolate SSH configuration reads, support limited Include processing, deduplicate aliases, and degrade gracefully when discovery fails.
- Preserve case-sensitive SSH alias matching while retaining case-insensitive matching for other profile types.

Documentation:
- Document SSH host discovery, configuration parsing limits, launch behavior, restore behavior, and the configuration toggle.
- Add SSH host discovery to the Windows capability documentation and status overview.

Tests:
- Add coverage for SSH profile construction, executable resolution, parser behavior, include limits, path rejection, launch hardening, session restore, split behavior, title handling, and working-directory precedence.

### CODE REVIEW

**Comment by amanthanvi** [comment]
@coderabbitai review
@codex review

Context for reviewers:

Parses `~/.ssh/config` into noctty profiles and the command palette. Activating an entry spawns `ssh.exe` via `Command.direct` argv with an explicit `lpApplicationName` — no `cmd.exe`, no shell string.

Prior adversarial review found and fixed two HIGH defects: (a) the **session-restore SSH guard** was bypassable by a case-mismatched key — `isSshProfileKey` used case-*sensitive* `startsWith(key, "ssh:")` while `profileIndexByKey` / `Host.profileForKey` matched with `ascii.eqlIgnoreCase`, so `"profile":"SSH:prod"` written into `session-state.json` slipped the guard but still resolved the real profile: unattended dial-out at next launch, and code execution given an attacker-controlled `ProxyCommand`. (b) Forced `shell-integration=none` was bypassed on the **split** path, which flattened the ssh argv into `cmd.exe /C` with no quoting.

**The guard now resolves the profile first and branches on `profile.kind == .ssh`.** If you are about to suggest replacing that with a string prefix comparison — even a case-insensitive one — please do not: the entire bug class was the guard's comparison semantics drifting from the lookup's. Branching on the resolved kind makes that drift structurally impossible.

Please scrutinise most:
1. **Alias validation** (`isConnectableAlias`) as an option-injection barrier — leading `-`, shell metacharacters, control bytes, and non-UTF-8 bytes reaching Win32 `DrawTextW` in the palette row.
2. **Include handling** in `windows_ssh_hosts.zig` — synchronous UI-thread reads, UNC/device paths in `Include`, the depth cap versus the uncapped include *count*, and the blast radius at session restore (this runs for the first pane of every restored window).
3. Whether an SSH-side parse failure can destroy the entire shell-profile list rather than degrading to `shell_profiles`.

File: src/config/windows_ssh_hosts.zig [hunk 1]
```diff
@@ -0,0 +1,549 @@
...
+            },
+        };
+        defer self.alloc.free(contents);
+        try parseInto(self.alloc, self.hosts, contents, self, depth);
+    }
+
+    fn parseIncludes(self: *LoadContext, value: []const u8, depth: u8) Allocator.Error!void {
+        if (depth >= 1) return;
+
+        var paths = std.mem.tokenizeAny(u8, value, " \t");
```
[the remark this answers is not in the document]
Comment (amanthanvi) [reply src/config/windows_ssh_hosts.zig]: **REAL — fixed.** Confirmed against the source at the PR head: `parseIncludes` split the directive value with `std.mem.tokenizeAny(u8, value, " \t")`, so `Include "C:\Users\Jane Doe\.ssh\work hosts"` became four tokens with the quote bytes still attached, every one of them a nonexistent path. `parseFile` maps `error.FileNotFound` to a silent `return`, so the failure mode was exactly as described: every host in that include silently omitted, no diagnostic.

One correction to the suggested approach. Routing `Include` through the existing `AliasIterator` would fix only half of it and break the other half:

- `AliasIterator.next` returns `self.value[start..self.index]` with the quote bytes still **in** the token. That is harmless on the `Host` path, where `isConnectableAlias` rejects `'` and `"` outright, but an include path has to be dequoted, not just delimited.
- `AliasIterator` treats `\` as an escape and skips the following byte. That is correct for `Host` patterns (and the existing `escaped\ alias` test depends on it) and actively wrong for a Windows path, where `\` is the separator.

So the fix is a small purpose-built tokenizer for include arguments rather than a reuse, matching OpenSSH `strdelim` semantics:

- `src/config/windows_ssh_hosts.zig:149` — `IncludeArgIterator`: whitespace-separated, a leading `"` groups one path containing spaces and is stripped, backslashes are copied verbatim, an unterminated quote consumes the remainder and yields nothing (OpenSSH rejects the line). Zero allocation — the yielded slice is a subslice of the directive value.
- `'` is deliberately **not** a grouping quote, matching OpenSSH, whose `QUOTE` set is `"` only. Treating it as one would corrupt a literal path such as `C:\Bob's\config`.
- `parseIncludes` now iterates it and skips empty tokens (`:102`).

Tests added:

- `"ssh config include tokenizer keeps quoted paths whole"` — the exact `Include "C:\Users\Jane Doe\.ssh\work hosts" plain.conf` case, backslash-verbatim, the `C:\Bob's\config` non-regression, unterminated quote, whitespace-only value.
- `"ssh config loads a quoted include whose filename contains spaces"` — end-to-end through `loadFromHome` against a real on-disk `.ssh/work hosts.conf`, asserting the host inside it now appears.

File: src/config/windows_ssh_hosts.zig [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
```
[the remark this answers is not in the document]
Comment (amanthanvi) [reply src/config/windows_ssh_hosts.zig]: **REAL — fixed.** Confirmed against the source at the PR head: `parseIncludes` split the directive value with `std.mem.tokenizeAny(u8, value, " \t")`, so `Include "C:\Users\Jane Doe\.ssh\work hosts"` became four tokens with the quote bytes still attached, every one of them a nonexistent path. `parseFile` maps `error.FileNotFound` to a silent `return`, so the failure mode was exactly as described: every host in that include silently omitted, no diagnostic.

One correction to the suggested approach. Routing `Include` through the existing `AliasIterator` would fix only half of it and break the other half:

- `AliasIterator.next` returns `self.value[start..self.index]` with the quote bytes still **in** the token. That is harmless on the `Host` path, where `isConnectableAlias` rejects `'` and `"` outright, but an include path has to be dequoted, not just delimited.
- `AliasIterator` treats `\` as an escape and skips the following byte. That is correct for `Host` patterns (and the existing `escaped\ alias` test depends on it) and actively wrong for a Windows path, where `\` is the separator.

So the fix is a small purpose-built tokenizer for include arguments rather than a reuse, matching OpenSSH `strdelim` semantics:

- `src/config/windows_ssh_hosts.zig:149` — `IncludeArgIterator`: whitespace-separated, a leading `"` groups one path containing spaces and is stripped, backslashes are copied verbatim, an unterminated quote consumes the remainder and yields nothing (OpenSSH rejects the line). Zero allocation — the yielded slice is a subslice of the directive value.
- `'` is deliberately **not** a grouping quote, matching OpenSSH, whose `QUOTE` set is `"` only. Treating it as one would corrupt a literal path such as `C:\Bob's\config`.
- `parseIncludes` now iterates it and skips empty tokens (`:102`).

Tests added:

- `"ssh config include tokenizer keeps quoted paths whole"` — the exact `Include "C:\Users\Jane Doe\.ssh\work hosts" plain.conf` case, backslash-verbatim, the `C:\Bob's\config` non-regression, unterminated quote, whitespace-only value.
- `"ssh config loads a quoted include whose filename contains spaces"` — end-to-end through `loadFromHome` against a real on-disk `.ssh/work hosts.conf`, asserting the host inside it now appears.

File: src/config/windows_ssh_hosts.zig [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+        while (paths.next()) |raw_path| {
```
[the remark this answers is not in the document]
Comment (amanthanvi) [reply src/config/windows_ssh_hosts.zig]: **REAL — fixed.** Confirmed against the source at the PR head: `parseIncludes` split the directive value with `std.mem.tokenizeAny(u8, value, " \t")`, so `Include "C:\Users\Jane Doe\.ssh\work hosts"` became four tokens with the quote bytes still attached, every one of them a nonexistent path. `parseFile` maps `error.FileNotFound` to a silent `return`, so the failure mode was exactly as described: every host in that include silently omitted, no diagnostic.

One correction to the suggested approach. Routing `Include` through the existing `AliasIterator` would fix only half of it and break the other half:

- `AliasIterator.next` returns `self.value[start..self.index]` with the quote bytes still **in** the token. That is harmless on the `Host` path, where `isConnectableAlias` rejects `'` and `"` outright, but an include path has to be dequoted, not just delimited.
- `AliasIterator` treats `\` as an escape and skips the following byte. That is correct for `Host` patterns (and the existing `escaped\ alias` test depends on it) and actively wrong for a Windows path, where `\` is the separator.

So the fix is a small purpose-built tokenizer for include arguments rather than a reuse, matching OpenSSH `strdelim` semantics:

- `src/config/windows_ssh_hosts.zig:149` — `IncludeArgIterator`: whitespace-separated, a leading `"` groups one path containing spaces and is stripped, backslashes are copied verbatim, an unterminated quote consumes the remainder and yields nothing (OpenSSH rejects the line). Zero allocation — the yielded slice is a subslice of the directive value.
- `'` is deliberately **not** a grouping quote, matching OpenSSH, whose `QUOTE` set is `"` only. Treating it as one would corrupt a literal path such as `C:\Bob's\config`.
- `parseIncludes` now iterates it and skips empty tokens (`:102`).

Tests added:

- `"ssh config include tokenizer keeps quoted paths whole"` — the exact `Include "C:\Users\Jane Doe\.ssh\work hosts" plain.conf` case, backslash-verbatim, the `C:\Bob's\config` non-regression, unterminated quote, whitespace-only value.
- `"ssh config loads a quoted include whose filename contains spaces"` — end-to-end through `loadFromHome` against a real on-disk `.ssh/work hosts.conf`, asserting the host inside it now appears.

File: src/config/windows_ssh_hosts.zig [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+                continue;
+            }
+            if (self.includes_read >= max_include_files) return;
+            self.includes_read += 1;
+
+            const path = if (std.mem.startsWith(u8, raw_path, "~/") or
+                std.mem.startsWith(u8, raw_path, "~\\"))
+                try std.fs.path.join(self.alloc, &.{ self.home_dir, raw_path[2..] })
+            else if (std.fs.path.isAbsolute(raw_path))
+                try self.alloc.dupe(u8, raw_path)
```
[the remark this answers is not in the document]
Comment (amanthanvi) [reply src/config/windows_ssh_hosts.zig:151]: **REAL — fixed.** Confirmed both halves against the source at the PR head.

The gate was `isRemoteOrDevicePath`, a pure syntax check for `\` / `//` prefixes. `Z:\ssh\hosts.conf` is syntactically an ordinary local absolute path, so it passed the gate, satisfied `std.fs.path.isAbsolute`, and reached `readFile` -> `std.fs.openFileAbsolute`.

The UI-thread claim is also correct, and the session-restore path is worse than the general case: `App.applyRestoredProfileConfig` (`src/apprt/win32.zig:2318`) calls `windows_shell.listProfiles(alloc, self.config."ssh-config-hosts")` synchronously whenever `host == null`, i.e. for the first pane of every restored window. An N-window restore multiplies the stall.

Fix — `src/config/windows_ssh_hosts.zig`:

- `driveLetter` (`:335`) — pure extraction of the `X` from an `X:\` / `X:/` root. Returns null for relative paths, UNC, `~/`, and the `X:name` drive-relative spelling.
- `isRemoteDriveLetter` (`:352`) — `GetDriveTypeW` on that root, rejecting `DRIVE_REMOTE`. This is the direct check and is genuinely non-blocking: `GetDriveTypeW` answers from the process's local mount table and does not contact the server, so it cannot itself hang on a dead mapping.
- `isUnsupportedIncludePath` (`:319`) = `isRemoteOrDevicePath or isRemoteDriveLetter`; `parseIncludes` now gates on that, still before `parseFile`.

On the include count: it was already capped. `max_include_files = 16` (`:10`) with `self.includes_read` enforced in `parseIncludes`, and depth capped at one level. So the "up to 16 serially" figure was the correct ceiling, and the fix removes the per-path stall rather than the count.

I did not skip the whole SSH parse on the session-restore path. With remote-backed includes rejected before they are opened, what remains is a few small local file reads, and skipping it would silently change which profile a restored window resolves. That the restore path re-enumerates profiles per window is a separate cost and out of scope here.

Tests added: `"ssh config recognizes drive-letter roots for the remote-drive check"` covers `driveLetter` across drive roots, forward-slash roots, relative, UNC, `~/`, and the `X:name` spelling, and asserts the local system drive is never misread as remote (which would drop every ordinary absolute include).

What remains, stated plainly: this covers a **mapped drive letter** only. A directory junction, or a `subst` alias whose target is remote, is a different vector and is not detected — both report the drive type of the letter rather than of the target. That limitation is now written into `docs/windows.md#ssh-hosts` rather than left implicit.

**Comment by amanthanvi** [comment]
@coderabbitai review
@codex review

Round 2, pushed as `0f91cc7`. Both round-1 findings were real and are fixed.

**Quoted `Include` paths — fixed.** `tokenizeAny` split the value on every space and tab with the quote bytes left in, and `parseFile` maps `FileNotFound` to a silent return, so every host in such an include vanished with no diagnostic at all.

I deliberately did **not** reuse the existing `Host`-line tokenizer, having checked that it would fix only half the problem: it yields tokens with the quote bytes still attached (harmless for `Host`, where `isConnectableAlias` rejects quotes outright, but wrong for a path that must be opened), and it treats `\` as an escape — correct for `Host` patterns, actively wrong for a Windows path separator. There is now a dedicated `IncludeArgIterator` matching OpenSSH's `strdelim`: a leading `"` groups and is stripped, backslashes are verbatim, an unterminated quote consumes the remainder, zero allocation. `'` is deliberately **not** a grouping quote, because OpenSSH's `QUOTE` is `"` only and stripping it would corrupt a path like `C:\Bob's\config`.

**Mapped-drive `Include` paths — fixed.** `isRemoteOrDevicePath` is pure syntax, so `Z:\...` passed it and reached `openFileAbsolute` on the UI thread. The blast radius was worse than first reported: `applyRestoredProfileConfig` calls `listProfiles` synchronously for the first pane of *every* restored window. Remote drive letters are now rejected before the open via `GetDriveTypeW` → `DRIVE_REMOTE`, which reads the local mount table and cannot itself block. The include *count* was already capped at 16 with depth 1, so that half of the earlier concern was already closed.

Stated honestly in the reply and in `docs/windows.md`: a junction or a `subst` alias pointing at a remote target is **not** covered, because both report the drive type of the letter. I did not skip SSH parsing on the session-restore path — with remote includes rejected pre-open what remains is a few small local reads, and skipping would silently change which profile a restored window resolves to.

The session-restore guard still branches on the resolved `profile.kind == .ssh`. **If you are about to suggest a string prefix comparison — even case-insensitive — please don't.** The original vulnerability was precisely the guard's comparison semantics drifting from the lookup's, which produced an unattended dial-out; branching on the resolved kind makes that drift structurally impossible.

Please concentrate on:
1. **`IncludeArgIterator` against real OpenSSH `strdelim` semantics** — embedded quotes, an unterminated quote, trailing separators, an empty argument, and multiple arguments on one `Include` line. I would rather find a parser divergence now than ship a path we silently fail to read.
2. **Whether any remaining `Include` path can still reach a blocking open** — device paths, `\\?\` forms, a relative path resolving onto a mapped drive, or a drive letter whose type changes between the check and the open.
3. **The `GetDriveTypeW` call itself** — argument form (it wants a root path with a trailing separator) and behaviour for a letter that is not mounted at all.

File: src/config/windows_ssh_hosts.zig [hunk 5]
```diff
@@ -0,0 +1,699 @@
...
+            },
+        };
+        defer self.alloc.free(contents);
+        try parseInto(self.alloc, self.hosts, contents, self, depth);
+    }
+
+    fn parseIncludes(self: *LoadContext, value: []const u8, depth: u8) Allocator.Error!void {
+        if (depth >= 1) return;
+
+        var paths: IncludeArgIterator = .{ .value = value };
```
[the remark this answers is not in the document]
Comment (amanthanvi) [reply src/config/windows_ssh_hosts.zig]: Declining — already fixed at the commit this was reported against (`0f91cc7`).

`parseIncludes` no longer tokenizes on whitespace. At `src/config/windows_ssh_hosts.zig:102` it reads:

```zig
var paths: IncludeArgIterator = .{ .value = value };
```

`IncludeArgIterator` is defined at `src/config/windows_ssh_hosts.zig:149`. It keeps a double-quoted argument whole, and deliberately does *not* treat `\` as an escape or `'` as a grouping quote, because both would corrupt Windows path separators and literal paths such as `C:\Bob's\config`. `std.mem.tokenizeAny` is gone from this path entirely — `grep tokenizeAny src/config/windows_ssh_hosts.zig` returns nothing at this head.

The requested regression test exists, plus the adjacent edge cases:

- `test "ssh config include tokenizer keeps quoted paths whole"` (`:541`) covers the quoted path with spaces (`"C:\Users\Jane Doe\.ssh\work hosts"` followed by a second unquoted argument), backslashes and apostrophes kept verbatim (`C:\Bob's\config`), an unterminated quote (consumed rather than guessed at, matching OpenSSH rejecting the line), and an all-whitespace value.
- `test "ssh config loads a quoted include whose filename contains spaces"` (`:562`) is the end-to-end version: it writes `.ssh/work hosts.conf` and asserts the alias inside it is loaded.

No code change on this one. If you are still seeing the whitespace split, please re-check against `0f91cc7` and point at the specific line, since `:102` there is the iterator construction rather than a tokenizer call.

**Comment by amanthanvi** [comment]
@coderabbitai review

@codex review

Final pre-merge review pass. This PR has already been through adversarial review and at least one bot round; findings were triaged as fix / decline-with-evidence / follow-up rather than applied blindly, so some flagged items are deliberate design with reasoning recorded in-thread. Please focus on anything that would be wrong *at merge time*: stale claims in the description versus the current head, cross-PR interactions, and correctness of security-relevant paths. If you flag something previously declined, please engage with the recorded reasoning rather than restating the original finding.

**Comment by amanthanvi** [comment]
@coderabbitai full review
@greptileai review this draft
@codex review

**Comment by amanthanvi** [comment]
Review fixes pushed in 2974679db.

Addressed:
- OpenSSH-like directive arguments now remove double quotes anywhere in a token, consume unterminated quoted remainders, preserve Windows backslashes for Include paths, and retain embedded/quoted `#` values while stripping only boundary comments.
- SSH session persistence now drops cwd as well as profile/title state; SSH split launches do not inherit a remote-controlled cwd.
- SSH title identity derives from persistent `launched_ssh` + `launch_profile_key`, so profile refresh/removal cannot strand the alias title.
- The generated config reference now discloses the bounded one-level, 16-file Include processing.

Validation observed:
- `zig build test -Dtest-filter="ssh config"`: 83 passed, 1 skipped.
- `zig build test -Dtest-filter="win32 ssh"`: 71 passed, 1 skipped.
- `zig build test -Demit-test-exe=true`: 3795 passed, 70 skipped.
- `zig build -Demit-exe=true`: 73/73 steps succeeded.
- `scripts/check-source-format.ps1` and `git diff --check`: passed.

The drive-type check/use race remains open intentionally: this branch's profile enumeration API is synchronous, and making arbitrary `CreateFile` opens bounded requires moving discovery behind an async/cache boundary rather than a safe narrow review patch. Please re-evaluate that thread against the current head.

**Comment by amanthanvi** [comment]
@coderabbitai review

**Comment by amanthanvi** [comment]
@greptileai

**Comment by amanthanvi** [comment]
@codex review

**Comment by amanthanvi** [comment]
The remaining Include-open race is fixed in 49d395cf0.

Include reads now run on a detached worker rather than the UI thread. Each refresh has one shared 50 ms Include budget, and a process-wide single-in-flight guard prevents a blocked network open from creating unbounded workers on later picker refreshes. The request is reference-counted and page-allocator-owned, so a timeout returns immediately while the worker retains and eventually frees its path/data/context. A later profile refresh can retry after the worker finishes.

Regression coverage exercises both a successful off-thread read and a deliberately slow read that times out at 1 ms, then verifies the single-flight guard recovers. Observed validation:
- `zig build test -Dtest-filter="ssh config"`: 84 passed, 1 skipped.
- `zig build test -Demit-test-exe=true`: 3796 passed, 70 skipped.
- source-format and `git diff --check`: passed.

**Comment by amanthanvi** [comment]
@coderabbitai review

**Comment by amanthanvi** [comment]
@greptileai

**Comment by amanthanvi** [comment]
@codex review

**Comment by amanthanvi** [comment]
Case-distinct SSH aliases are preserved in d8b2fc177.

- Discovery now deduplicates only exact repeats, so `PROD` and `prod` both survive.
- Live SSH profile-key, exact picker, preferred-profile, split, and quick-slot lookup is exact-case; local Windows profile keys retain their previous case-insensitive behavior.
- Session restore still refuses every case variant of an `ssh:` key before lookup, so the safety gate did not regress.

Observed validation:
- `ssh config`: 84 passed, 1 skipped.
- `profileIndexByKey`: 68 passed, 1 skipped.
- `session restore refuses ssh`: 68 passed, 1 skipped.
- full emitted suite: 3796 passed, 70 skipped.
- source-format and `git diff --check`: passed.

**Comment by amanthanvi** [comment]
@coderabbitai review

**Comment by amanthanvi** [comment]
@greptileai

**Comment by amanthanvi** [comment]
@codex review

**Comment by amanthanvi** [comment]
Addressed the two fresh Codex findings in `b3c69162e`:

- SSH profile caches are invalidated whenever `ssh-config-hosts` changes. An open profile picker reloads immediately; the universal palette and other consumers reload lazily through the existing `ensureProfiles()` seam, so stale SSH entries cannot remain launchable after disabling the setting.
- The root `%USERPROFILE%\.ssh\config` open now uses the same detached, single-in-flight bounded read worker as Include files. Root plus Include reads share one 50 ms refresh budget, and the docs now state that contract.

Validation at this head:
- `scripts\dev-windows.cmd zig build test '-Dtest-filter=ssh config' --summary all` — PASS, 84 passed / 1 skipped.
- `scripts\dev-windows.cmd zig build test -Demit-test-exe=true --summary all` — PASS, 3,796 passed / 70 skipped.
- `pwsh -NoProfile -File scripts/check-source-format.ps1` — PASS.
- `git diff --check` — PASS.

**Comment by amanthanvi** [comment]
@coderabbitai review

**Comment by amanthanvi** [comment]
@greptileai

**Comment by amanthanvi** [comment]
@codex review

**Comment by amanthanvi** [comment]
@coderabbitai review

**Comment by amanthanvi** [comment]
Fixed in `9f8d4ff4a`. `readFileWithinBudget` now returns an explicit result that distinguishes a missing path, another skippable read failure, and exhausted timeout budget. `parseIncludes` continues past missing/skipped arguments while still stopping when the shared 50 ms budget is exhausted; the root config path retains its empty-list behavior for missing/unavailable files.

Added an end-to-end regression for `Include missing.conf valid.conf`, asserting both the valid included alias and the root alias survive.

Validation:
- `scripts\dev-windows.cmd zig build test '-Dtest-filter=ssh config' --summary all` — PASS, 85 passed / 1 skipped.
- `scripts\dev-windows.cmd zig build test -Demit-test-exe=true --summary all` — PASS, 3,797 passed / 70 skipped.
- `pwsh -NoProfile -File scripts/check-source-format.ps1` — PASS.
- `git diff --check` — PASS.

**Comment by amanthanvi** [comment]
@coderabbitai review

**Comment by amanthanvi** [comment]
@greptileai

**Comment by amanthanvi** [comment]
@codex review

**Comment by amanthanvi** [comment]
Updated the PR description to match the reviewed behavior: SSH aliases are deduplicated only on exact byte equality, so case-distinct OpenSSH aliases such as `prod` and `PROD` remain distinct. I also removed two now-fixed residuals and updated the session/include descriptions to the current exact-lookup and bounded-worker contracts.

**Comment by amanthanvi** [comment]
Fixed in `a387368f2`. The obsolete synchronous/UI-thread claim is gone. The paragraph now states that known globs/UNC/device/mapped-drive paths are rejected before the worker opens them, while undetected junction/`subst` targets remain protected by the bounded worker described below. `scripts/check-source-format.ps1` and `git diff --check` pass; no production code changed.

**Comment by amanthanvi** [comment]
@coderabbitai review

**Comment by amanthanvi** [comment]
@greptileai

**Comment by amanthanvi** [comment]
@codex review

**Comment by amanthanvi** [comment]
Fixed in rebased commit `4422e2e37`. The parser now treats the first `Host` or `Match` directive as the start of conditional context and harvests `Include` aliases only before that point. Includes inside either conditional block are skipped, while each included file still starts with its own global context.

The existing end-to-end include test now adds `Host scoped` followed by `Include host-scoped.conf` and proves the included `Host leaked` alias is not exposed, while later concrete `Host root` remains discoverable.

Rebase/validation evidence:
- Rebasing onto `origin/main` at `bfcec8789` had one documentation-table conflict; resolution preserved both main's power-aware row and this PR's SSH-discovery row.
- `scripts\dev-windows.cmd zig build test '-Dtest-filter=ssh config' --summary all` — PASS, 85 passed / 1 skipped.
- `scripts\dev-windows.cmd zig build test -Demit-test-exe=true --summary all` — PASS, 3,811 passed / 70 skipped.
- `pwsh -NoProfile -File scripts/check-source-format.ps1` — PASS.
- `git diff --check` — PASS.
- Rebasing force-push used `--force-with-lease`.

**Comment by amanthanvi** [comment]
@coderabbitai review

**Comment by amanthanvi** [comment]
@greptileai

**Comment by amanthanvi** [comment]
@codex review
